File: sdk/typescript/src/namespaces/ar-automation.ts
/**
 * AR Automation Namespace API
 *
 * Provides Accounts Receivable automation:
 * - Invoice creation and sending
 * - Payment reminders
 * - Aging reports
 * - Collection suggestions
 * - Customer balance tracking
 */

/**
 * AR Invoice status.
 */
export type ARInvoiceStatus = 'draft' | 'sent' | 'viewed' | 'paid' | 'partial' | 'overdue';

/**
 * Reminder escalation level.
 */
export type ReminderLevel = 1 | 2 | 3 | 4;

/**
 * Line item on an invoice.
 */
export interface ARLineItem {
  description: string;
  quantity?: number;
  unit_price?: number;
  amount: number;
  tax_rate?: number;
}

/**
 * AR Invoice.
 */
export interface ARInvoice {
  id: string;
  customer_id: string;
  customer_name: string;
  customer_email?: string;
  invoice_number: string;
  invoice_date: string;
  due_date: string;
  total_amount: number;
  tax_amount?: number;
  amount_paid: number;
  balance_due: number;
  payment_terms: string;
  status: ARInvoiceStatus;
  line_items: ARLineItem[];
  notes?: string;
  last_reminder_sent?: string;
  reminder_count: number;
  created_at: string;
  updated_at?: string;
}

/**
 * Create AR invoice request.
 */
export interface CreateARInvoiceRequest {
  customer_id: string;
  customer_name: string;
  customer_email?: string;
  invoice_date?: string;
  due_date: string;
  payment_terms?: string;
  line_items: ARLineItem[];
  notes?: string;
  send_immediately?: boolean;
}

/**
 * List AR invoices parameters.
 */
export interface ListARInvoicesParams {
  limit?: number;
  offset?: number;
  status?: ARInvoiceStatus;
  customer_id?: string;
  overdue_only?: boolean;
  due_before?: string;
  due_after?: string;
}

/**
 * Record AR payment request.
 */
export interface RecordARPaymentRequest {
  amount: number;
  payment_date?: string;
  payment_method?: string;
  reference_number?: string;
  notes?: string;
}

/**
 * Aging bucket.
 */
export interface AgingBucket {
  label: string;
  days_range: string;
  count: number;
  total_amount: number;
  invoices: Array<{
    invoice_id: string;
    customer_name: string;
    amount: number;
    days_overdue: number;
  }>;
}

/**
 * Aging report.
 */
export interface AgingReport {
  as_of_date: string;
  total_outstanding: number;
  total_overdue: number;
  buckets: {
    current: AgingBucket;
    days_1_30: AgingBucket;
    days_31_60: AgingBucket;
    days_61_90: AgingBucket;
    days_over_90: AgingBucket;
  };
  generated_at: string;
}

/**
 * Collection action type.
 */
export type CollectionActionType =
  | 'send_reminder'
  | 'phone_call'
  | 'escalate_to_collections'
  | 'offer_payment_plan'
  | 'final_notice';

/**
 * Collection suggestion.
 */
export interface CollectionSuggestion {
  invoice_id: string;
  customer_id: string;
  customer_name: string;
  amount_due: number;
  days_overdue: number;
  suggested_action: CollectionActionType;
  action_priority: 'high' | 'medium' | 'low';
  reason: string;
  previous_attempts: number;
}

/**
 * Add customer request.
 */
export interface AddARCustomerRequest {
  name: string;
  email?: string;
  phone?: string;
  billing_address?: {
    street?: string;
    city?: string;
    state?: string;
    zip?: string;
    country?: string;
  };
  payment_terms?: string;
  credit_limit?: number;
  notes?: string;
}

/**
 * Customer balance.
 */
export interface CustomerBalance {
  customer_id: string;
  customer_name: string;
  total_outstanding: number;
  total_overdue: number;
  invoice_count: number;
  oldest_invoice_date?: string;
  average_days_to_pay?: number;
  credit_limit?: number;
  available_credit?: number;
}

/**
 * Client interface for AR automation operations.
 */
interface ARAutomationClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }
  ): Promise<T>;
}

/**
 * AR Automation API namespace.
 *
 * Provides methods for Accounts Receivable automation:
 * - Create and send invoices
 * - Send payment reminders
 * - Generate aging reports
 * - Get collection suggestions
 * - Track customer balances
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai' });
 *
 * // Create and send an invoice
 * const invoice = await client.arAutomation.createInvoice({
 *   customer_id: 'cust_123',
 *   customer_name: 'Acme Corp',
 *   customer_email: 'billing@acme.com',
 *   due_date: '2025-02-28',
 *   line_items: [{ description: 'Consulting', amount: 5000 }],
 *   send_immediately: true,
 * });
 *
 * // Get aging report
 * const aging = await client.arAutomation.getAgingReport();
 * ```
 */
export class ARAutomationAPI {
  constructor(private client: ARAutomationClientInterface) {}

  /**
   * Create a new AR invoice.
   */
  async createInvoice(
    request: CreateARInvoiceRequest
  ): Promise<{ invoice: ARInvoice; message: string }> {
    return this.client.request('POST', '/api/v1/accounting/ar/invoices', {
      json: request,
    });
  }

  /**
   * List AR invoices with filtering.
   */
  async listInvoices(
    params?: ListARInvoicesParams
  ): Promise<{ invoices: ARInvoice[]; total: number; limit: number; offset: number }> {
    return this.client.request('GET', '/api/v1/accounting/ar/invoices', {
      params: params as Record<string, unknown>,
    });
  }

  /**
   * Get a specific AR invoice.
   */
  async getInvoice(invoiceId: string): Promise<{ invoice: ARInvoice }> {
    return this.client.request('GET', `/api/v1/accounting/ar/invoices/${invoiceId}`);
  }

  /**
   * Send an invoice to the customer.
   */
  async sendInvoice(invoiceId: string): Promise<{ success: boolean; message: string }> {
    return this.client.request('POST', `/api/v1/accounting/ar/invoices/${invoiceId}/send`);
  }

  /**
   * Send a payment reminder for an invoice.
   */
  async sendReminder(
    invoiceId: string,
    escalationLevel?: ReminderLevel
  ): Promise<{ success: boolean; message: string; next_escalation?: ReminderLevel }> {
    return this.client.request('POST', `/api/v1/accounting/ar/invoices/${invoiceId}/reminder`, {
      json: escalationLevel ? { escalation_level: escalationLevel } : undefined,
    });
  }

  /**
   * Record a payment against an invoice.
   */
  async recordPayment(
    invoiceId: string,
    request: RecordARPaymentRequest
  ): Promise<{ invoice: ARInvoice; message: string }> {
    return this.client.request('POST', `/api/v1/accounting/ar/invoices/${invoiceId}/payment`, {
      json: request,
    });
  }

  /**
   * Get AR aging report.
   */
  async getAgingReport(): Promise<{ aging_report: AgingReport }> {
    return this.client.request('GET', '/api/v1/accounting/ar/aging');
  }

  /**
   * Get collection suggestions for overdue invoices.
   */
  async getCollectionSuggestions(): Promise<{ suggestions: CollectionSuggestion[] }> {
    return this.client.request('GET', '/api/v1/accounting/ar/collections');
  }

  /**
   * Add a new customer.
   */
  async addCustomer(
    request: AddARCustomerRequest
  ): Promise<{ success: boolean; customer_id: string; message: string }> {
    return this.client.request('POST', '/api/v1/accounting/ar/customers', {
      json: request,
    });
  }

  /**
   * Get customer balance summary.
   */
  async getCustomerBalance(customerId: string): Promise<CustomerBalance> {
    return this.client.request('GET', `/api/v1/accounting/ar/customers/${customerId}/balance`);
  }
}

File: sdk/python/aragora/namespaces/ar_automation.py
"""
AR Automation Namespace API.

Provides Accounts Receivable automation:
- Invoice creation and sending
- Payment reminders
- Aging reports
- Collection suggestions
- Customer balance tracking
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient

ARInvoiceStatus = Literal["draft", "sent", "viewed", "paid", "partial", "overdue"]
ReminderLevel = Literal[1, 2, 3, 4]
CollectionActionType = Literal[
    "send_reminder", "phone_call", "escalate_to_collections", "offer_payment_plan", "final_notice"
]


class ARAutomationAPI:
    """
    Synchronous Accounts Receivable Automation API.

    Provides methods for:
    - Invoice creation and sending
    - Payment reminders and escalation
    - Aging reports and analytics
    - Collection suggestions
    - Customer balance tracking
    """

    def __init__(self, client: AragoraClient) -> None:
        self._client = client

    # =========================================================================
    # Invoice Management
    # =========================================================================

    def create_invoice(
        self,
        customer_id: str,
        customer_name: str,
        due_date: str,
        line_items: list[dict[str, Any]],
        customer_email: str | None = None,
        invoice_date: str | None = None,
        payment_terms: str | None = None,
        notes: str | None = None,
        send_immediately: bool = False,
    ) -> dict[str, Any]:
        """
        Create a new AR invoice.

        Args:
            customer_id: Customer identifier.
            customer_name: Customer display name.
            due_date: Payment due date (ISO format).
            line_items: List of line items with description and amount.
            customer_email: Customer email for sending invoice.
            invoice_date: Invoice date (ISO format).
            payment_terms: Payment terms (e.g., "Net 30").
            notes: Additional notes.
            send_immediately: Send invoice immediately after creation.

        Returns:
            Created invoice details.
        """
        data: dict[str, Any] = {
            "customer_id": customer_id,
            "customer_name": customer_name,
            "due_date": due_date,
            "line_items": line_items,
        }
        if customer_email:
            data["customer_email"] = customer_email
        if invoice_date:
            data["invoice_date"] = invoice_date
        if payment_terms:
            data["payment_terms"] = payment_terms
        if notes:
            data["notes"] = notes
        if send_immediately:
            data["send_immediately"] = send_immediately

        return self._client._request("POST", "/api/v1/accounting/ar/invoices", json=data)

    def list_invoices(
        self,
        status: ARInvoiceStatus | None = None,
        customer_id: str | None = None,
        overdue_only: bool = False,
        due_before: str | None = None,
        due_after: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List AR invoices with filtering.

        Args:
            status: Filter by invoice status.
            customer_id: Filter by customer.
            overdue_only: Only show overdue invoices.
            due_before: Filter by due date (before).
            due_after: Filter by due date (after).
            limit: Maximum results.
            offset: Pagination offset.

        Returns:
            List of invoices with total count.
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if status:
            params["status"] = status
        if customer_id:
            params["customer_id"] = customer_id
        if overdue_only:
            params["overdue_only"] = overdue_only
        if due_before:
            params["due_before"] = due_before
        if due_after:
            params["due_after"] = due_after

        return self._client._request("GET", "/api/v1/accounting/ar/invoices", params=params)

    def get_invoice(self, invoice_id: str) -> dict[str, Any]:
        """
        Get AR invoice details.

        Args:
            invoice_id: Invoice identifier.

        Returns:
            Invoice details.
        """
        return self._client._request("GET", f"/api/v1/accounting/ar/invoices/{invoice_id}")

    def send_invoice(self, invoice_id: str) -> dict[str, Any]:
        """
        Send an invoice to the customer.

        Args:
            invoice_id: Invoice identifier.

        Returns:
            Send confirmation.
        """
        return self._client._request("POST", f"/api/v1/accounting/ar/invoices/{invoice_id}/send")

    def send_reminder(
        self,
        invoice_id: str,
        escalation_level: ReminderLevel | None = None,
    ) -> dict[str, Any]:
        """
        Send a payment reminder for an invoice.

        Args:
            invoice_id: Invoice identifier.
            escalation_level: Reminder escalation level (1-4).

        Returns:
            Reminder confirmation with next escalation level.
        """
        data: dict[str, Any] = {}
        if escalation_level:
            data["escalation_level"] = escalation_level

        return self._client._request(
            "POST", f"/api/v1/accounting/ar/invoices/{invoice_id}/reminder", json=data
        )

    def record_payment(
        self,
        invoice_id: str,
        amount: float,
        payment_date: str | None = None,
        payment_method: str | None = None,
        reference_number: str | None = None,
        notes: str | None = None,
    ) -> dict[str, Any]:
        """
        Record a payment against an AR invoice.

        Args:
            invoice_id: Invoice identifier.
            amount: Payment amount.
            payment_date: Payment date (ISO format).
            payment_method: Payment method used.
            reference_number: Check/reference number.
            notes: Payment notes.

        Returns:
            Payment confirmation with updated invoice.
        """
        data: dict[str, Any] = {"amount": amount}
        if payment_date:
            data["payment_date"] = payment_date
        if payment_method:
            data["payment_method"] = payment_method
        if reference_number:
            data["reference_number"] = reference_number
        if notes:
            data["notes"] = notes

        return self._client._request(
            "POST", f"/api/v1/accounting/ar/invoices/{invoice_id}/payment", json=data
        )

    # =========================================================================
    # Aging and Collections
    # =========================================================================

    def get_aging_report(self) -> dict[str, Any]:
        """
        Get AR aging report.

        Returns:
            Aging report with buckets (current, 1-30, 31-60, 61-90, 90+).
        """
        return self._client._request("GET", "/api/v1/accounting/ar/aging")

    def get_collection_suggestions(self) -> dict[str, Any]:
        """
        Get collection suggestions for overdue invoices.

        Returns:
            List of suggested collection actions.
        """
        return self._client._request("GET", "/api/v1/accounting/ar/collections")

    # =========================================================================
    # Customer Management
    # =========================================================================

    def add_customer(
        self,
        name: str,
        email: str | None = None,
        phone: str | None = None,
        billing_address: dict[str, str] | None = None,
        payment_terms: str | None = None,
        credit_limit: float | None = None,
        notes: str | None = None,
    ) -> dict[str, Any]:
        """
        Add a new customer.

        Args:
            name: Customer name.
            email: Customer email.
            phone: Customer phone.
            billing_address: Billing address dict (street, city, state, zip, country).
            payment_terms: Default payment terms.
            credit_limit: Credit limit.
            notes: Additional notes.

        Returns:
            Created customer details.
        """
        data: dict[str, Any] = {"name": name}
        if email:
            data["email"] = email
        if phone:
            data["phone"] = phone
        if billing_address:
            data["billing_address"] = billing_address
        if payment_terms:
            data["payment_terms"] = payment_terms
        if credit_limit is not None:
            data["credit_limit"] = credit_limit
        if notes:
            data["notes"] = notes

        return self._client._request("POST", "/api/v1/accounting/ar/customers", json=data)

    def get_customer_balance(self, customer_id: str) -> dict[str, Any]:
        """
        Get customer balance summary.

        Args:
            customer_id: Customer identifier.

        Returns:
            Customer balance with outstanding and overdue amounts.
        """
        return self._client._request(
            "GET", f"/api/v1/accounting/ar/customers/{customer_id}/balance"
        )


class AsyncARAutomationAPI:
    """Asynchronous Accounts Receivable Automation API."""

    def __init__(self, client: AragoraAsyncClient) -> None:
        self._client = client

    # =========================================================================
    # Invoice Management
    # =========================================================================

    async def create_invoice(
        self,
        customer_id: str,
        customer_name: str,
        due_date: str,
        line_items: list[dict[str, Any]],
        customer_email: str | None = None,
        invoice_date: str | None = None,
        payment_terms: str | None = None,
        notes: str | None = None,
        send_immediately: bool = False,
    ) -> dict[str, Any]:
        """Create a new AR invoice."""
        data: dict[str, Any] = {
            "customer_id": customer_id,
            "customer_name": customer_name,
            "due_date": due_date,
            "line_items": line_items,
        }
        if customer_email:
            data["customer_email"] = customer_email
        if invoice_date:
            data["invoice_date"] = invoice_date
        if payment_terms:
            data["payment_terms"] = payment_terms
        if notes:
            data["notes"] = notes
        if send_immediately:
            data["send_immediately"] = send_immediately

        return await self._client._request("POST", "/api/v1/accounting/ar/invoices", json=data)

    async def list_invoices(
        self,
        status: ARInvoiceStatus | None = None,
        customer_id: str | None = None,
        overdue_only: bool = False,
        due_before: str | None = None,
        due_after: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List AR invoices with filtering."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if status:
            params["status"] = status
        if customer_id:
            params["customer_id"] = customer_id
        if overdue_only:
            params["overdue_only"] = overdue_only
        if due_before:
            params["due_before"] = due_before
        if due_after:
            params["due_after"] = due_after

        return await self._client._request("GET", "/api/v1/accounting/ar/invoices", params=params)

    async def get_invoice(self, invoice_id: str) -> dict[str, Any]:
        """Get AR invoice details."""
        return await self._client._request("GET", f"/api/v1/accounting/ar/invoices/{invoice_id}")

    async def send_invoice(self, invoice_id: str) -> dict[str, Any]:
        """Send an invoice to the customer."""
        return await self._client._request(
            "POST", f"/api/v1/accounting/ar/invoices/{invoice_id}/send"
        )

    async def send_reminder(
        self,
        invoice_id: str,
        escalation_level: ReminderLevel | None = None,
    ) -> dict[str, Any]:
        """Send a payment reminder for an invoice."""
        data: dict[str, Any] = {}
        if escalation_level:
            data["escalation_level"] = escalation_level

        return await self._client._request(
            "POST", f"/api/v1/accounting/ar/invoices/{invoice_id}/reminder", json=data
        )

    async def record_payment(
        self,
        invoice_id: str,
        amount: float,
        payment_date: str | None = None,
        payment_method: str | None = None,
        reference_number: str | None = None,
        notes: str | None = None,
    ) -> dict[str, Any]:
        """Record a payment against an AR invoice."""
        data: dict[str, Any] = {"amount": amount}
        if payment_date:
            data["payment_date"] = payment_date
        if payment_method:
            data["payment_method"] = payment_method
        if reference_number:
            data["reference_number"] = reference_number
        if notes:
            data["notes"] = notes

        return await self._client._request(
            "POST", f"/api/v1/accounting/ar/invoices/{invoice_id}/payment", json=data
        )

    # =========================================================================
    # Aging and Collections
    # =========================================================================

    async def get_aging_report(self) -> dict[str, Any]:
        """Get AR aging report."""
        return await self._client._request("GET", "/api/v1/accounting/ar/aging")

    async def get_collection_suggestions(self) -> dict[str, Any]:
        """Get collection suggestions for overdue invoices."""
        return await self._client._request("GET", "/api/v1/accounting/ar/collections")

    # =========================================================================
    # Customer Management
    # =========================================================================

    async def add_customer(
        self,
        name: str,
        email: str | None = None,
        phone: str | None = None,
        billing_address: dict[str, str] | None = None,
        payment_terms: str | None = None,
        credit_limit: float | None = None,
        notes: str | None = None,
    ) -> dict[str, Any]:
        """Add a new customer."""
        data: dict[str, Any] = {"name": name}
        if email:
            data["email"] = email
        if phone:
            data["phone"] = phone
        if billing_address:
            data["billing_address"] = billing_address
        if payment_terms:
            data["payment_terms"] = payment_terms
        if credit_limit is not None:
            data["credit_limit"] = credit_limit
        if notes:
            data["notes"] = notes

        return await self._client._request("POST", "/api/v1/accounting/ar/customers", json=data)

    async def get_customer_balance(self, customer_id: str) -> dict[str, Any]:
        """Get customer balance summary."""
        return await self._client._request(
            "GET", f"/api/v1/accounting/ar/customers/{customer_id}/balance"
        )
